### DISCUSSION TITLE
Ubuntu + WSL2 specific, beginner friendly build guide

### DISCUSSION THREAD

**Discussion by somekid6** [NoahGWood/OpenEFT#27] [General]
This guide assumes you have [a fresh copy of Ubuntu in WSL2 installed](https://ubuntu.com/tutorials/install-ubuntu-on-wsl2-on-windows-11-with-gui-support#1-overview) and is intended for a non-Ubuntu/WSL user audience.

I've compiled everyone's work so far and tried to make it as beginner friendly as I can. I removed the unGoogled Chrome bits since a working browser is installed by default on Windows usually anyways. I'm not a developer so this might not have the best practices but it does seem to work. I tested it on an install of Ubuntu on WSL2. It's got a few small changes compared to dedicated Ubuntu install  but they're small.

Open your fancy WSL2 Ubuntu install.

type 

	sudo apt-get update -y && sudo apt-get upgrade -y

 and press enter.

type 

	sudo apt install -y git cmake python3-pip libx11-dev

 and press enter.

type 

	cd ~/

and press enter.

type 

	git clone --recurse-submodules https://github.com/NoahGWood/OpenEFT.git

and press enter.

type

	cd OpenEFT
	
and press enter.

Open 'File Explorer' (this is the folder icon or you can use the 'Computer' icon) On the left side where it says 'This PC' scroll down until you see one that says 'Linux'. Click that once. Now follow the folders in this order: Ubuntu -> home -> <your username here> -> OpenEFT

Now double click 'requirements.txt' and it should open in Notepad.

Change:

	backports.zoneinfo==0.2.1
	
To:

	backports.zoneinfo;python_version<"3.9"
	
Save it and close Notepad.

Now go back to your files and double click "build_windows.sh". It will ask you how to open it (I recommend Notepad++ but Windows Notepad works fine). Click 'Notepad' from the options and 'Just Once' for now. 

Copy the script below and replace everything in the file with it, then save/close the file.
	

	#!/usr/bin/bash
	
	mkdir build
	
	echo "Setting Up NBIS"
	sudo rm -rf ./nbis #Remove any old copies
	git clone https://github.com/OwenPemberton/nbis ./nbis
	cd nbis
	./setup.sh ${PWD}/../build --64
	echo "Configuring NBIS"
	make config
	echo "Making NBIS"
	make it
	echo "Installing NBIS"
	sudo make install LIBNBIS=no
	
	echo "Checking NBIS binaries paths"
	case :$PATH:
	  in *:$HOME/OpenEFT/nbis/nfseg/bin:*) ;; # do nothing, it's there
	     *) echo 'export PATH="$HOME/OpenEFT/nbis/nfseg/bin:$PATH"' >> ~/.bashrc #it's missing, add to PATH
	esac
 
	case :$PATH:
	  in *:$HOME/OpenEFT/nbis/nfiq/bin:*) ;; # do nothing, it's there
	     *) echo 'export PATH="$HOME/OpenEFT/nbis/nfiq/bin:$PATH"' >> ~/.bashrc #it's missing, add to PATH
	esac

	#source ~/.bashrc doesn't work in .sh in Ubuntu since scripts run in their own session. Need to run post build
	
	echo "Installing LibOpenJP2-Tools"
	sudo apt-get install libopenjp2-tools -y
	sudo apt-get install libgl1-mesa-glx -y
	
	echo "Installing Pip Requirements"
	
	cd ~/OpenEFT
	pip3 install -r requirements.txt
	python3 manage.py migrate
	
	echo "OpenEFT build complete. Update your PATH by typing 'source ~/.bashrc' and run the app by typing 'python3 openeft.py'"
	cd ~/OpenEFT

Go back to your terminal window and type 

	cd ~/OpenEFT

 and press enter.

Now type 

	./build_windows.sh

and press enter. Input your password when it asks. 

Once it's done, type 

	source ~/.bashrc

 and press enter. 

Now type 

	python3 openeft.py

 and press enter to start the program. Ignore the error. It's trying to find chrome but it's not important.

Open your browser of choice, then put this in the URL bar

	localhost:7100

Press enter and you should be good to go!

edit - Did some testing and it looks like OpenJPEG isn't needed since moving libopenjp2-tools out of the if statement from the original so I removed that.

edit - added libgl1-mesa-glx dependency

edit - the eForm system will be unhappy if you submit the eft on the same day you made it. just wait a day and it'll be happy again.

**Comment by rosavicious** [comment]
Thank you!!! Got OPENEFT installed and almost working completely. It generates an EFT file but when I validate it through the NIST viewer it looks as if I am having issues with only the right/left four and right/left thumbs being used repeatedly...can't thank you enough for the effort you have put into this project...any help would be greatly appreciated...

![eftview](https://github.com/NoahGWood/OpenEFT/assets/79100128/8dd10563-9a97-4001-9b35-32131ef6e09e)

**Reply by ALeppala** [reply]
@somekid6 Awesome instructions!  Should have looked in here first.  Would have saved me hours of fiddle-farting around.  @rosavicious I had the same issue.  Only 3 images were showing in my generated eft file.  I compared it with one I had generated at the UPS Store.  That one had 14 images.  To make the results the same, I edited the \OpenEFT\conversion\core\fd258_ocr.py file.  Below are the contents of the file that will make it match the UPS Store version:

```
from collections import namedtuple

OCRLocation = namedtuple("OCRLocation", ["id", "bbox", "fp_number"])
OCR_LOCATIONS = [ 
# Not right locations #    OCRLocation("last_name", (7248,668,2581,468),0),
# Not right locations #    OCRLocation("first_name", (12674,668,2581,468),0),
# Not right locations #    OCRLocation("middle_name", (9954,668,2581,468),0),
# Not right locations #    OCRLocation("signature", (60,1770,6904,354),0),
# Not right locations #    OCRLocation("date", (60,3710,594,354),0),
# Not right locations #    OCRLocation("residence", (60,1184,3488,468),0),
# Not right locations #    OCRLocation("reason", (60,6045,6879,468),0),
# Not right locations #    OCRLocation("aliases", (7307,1794,3046,1152),0),
# Not right locations #    OCRLocation("citizenship", (7242,3318,3516,328),0),
# Not right locations #    OCRLocation("ssn", (7242,5950,3516,328),0),
# Not right locations #    OCRLocation("dob", (16266,2710,2832,288),0),
# Not right locations #    OCRLocation("pob", (16270,3338,2824,348),0),
# Not right locations #    OCRLocation("sex", (10842,3260,655,429),0),
# Not right locations #    OCRLocation("race", (11582,3338,608,348),0),
# Not right locations #    OCRLocation("height", (11582,3338,1072,348),0),
# Not right locations #    OCRLocation("weight", (11582,3338,772,348),0),
# Not right locations #    OCRLocation("eyes", (11582,3338,836,348),0),
# Not right locations #    OCRLocation("hair", (11582,3338,760,348),0),
    OCRLocation("R_THUMB", (90,7103,3700,3580), 1),
    OCRLocation("R_INDEX", (3884,7103,3835,3580), 2),
    OCRLocation("R_MIDDLE", (7753,7103,3796,3580), 3),
    OCRLocation("R_RING", (11583,7103,3836,3580), 4),
    OCRLocation("R_LITTLE", (15453,7103,3658,3580), 5),
    OCRLocation("L_THUMB", (90,10717,3700,3580), 6),
    OCRLocation("L_INDEX", (3884,10717,3835,3580), 7),
    OCRLocation("L_MIDDLE", (7753,10717,3796,3580), 8),
    OCRLocation("L_RING", (11583,10717,3836,3580), 9),
    OCRLocation("L_LITTLE", (15453,10717,3658,3580), 10),
    OCRLocation("R_THUMB2", (9653,14317,1866,4634), 11),
    OCRLocation("L_THUMB2", (7753,14317,1866,4634), 12),
    OCRLocation("R_FOUR", (11583,14317,7559,4634), 13),
    OCRLocation("L_FOUR", (90,14417,7559,4634), 14),
    #OCRLocation("2_THUMBS", (7753,14317,3795,4634), 15), # Consolidate into 1 record
]
```

I then saved and ran the build script again...  All good now!

**Reply by cmarti157** [reply]
I'm still getting 3 after changing this, wonder what I'm doing wrong. Hmm

**Comment by rosavicious** [comment]
Thank you so much!!!

**Comment by LuckyDucky87** [comment]
Edit1:  I see in another thread [Noah Wood commented](https://github.com/NoahGWood/OpenEFT/discussions/18#discussioncomment-5539041) that ATF eForms only supports the 4fingers and 2 thumbs pics anyway. Does this mean I don't need all 15 print pics?

Original post: 

@rosavicious
@ALeppala
@somekid6

Any ideas?

If I use the fd258_ocr.py file that comes with the original repo, it only saves 3 pictures to the eft (I can modify it to pick any 3, but only 3 nonetheless). If I uncomment out more than 3, it processes a few fingers then gives this error and no eft file.

CalledProcessError at /new/step2
Command 'nfseg 11 1 1 1 0 /tmp/tmp5s3d1z3a/R_THUMB2.jp2' returned non-zero exit status 223.
Request Method: POST
Request URL: http://localhost:7100/new/step2
Django Version: 4.1.7
Exception Type: CalledProcessError
Exception Value:
Command 'nfseg 11 1 1 1 0 /tmp/tmp5s3d1z3a/R_THUMB2.jp2' returned non-zero exit status 223.
Exception Location: /usr/lib/python3.10/subprocess.py, line 526, in run
Raised during: conversion.views.step2
Python Executable: /usr/bin/python3
Python Version: 3.10.12
Python Path:
['/home/Luckyducky87/OpenEFT',
'/usr/lib/python310.zip',
'/usr/lib/python3.10',
'/usr/lib/python3.10/lib-dynload',
'/home/Luckyducky87/.local/lib/python3.10/site-packages',
'/usr/local/lib/python3.10/dist-packages',
'/usr/lib/python3/dist-packages']

And the command line says "ERROR: putc_bytes : buffer overlow : alloc = 0, request = 2"

**Reply by ALeppala** [reply]
@LuckyDucky87 Did you follow the guide above by  @somekid6  to get it running?  I was getting a very similar error as well until I followed the steps exactly.  
MY issue now is that the EFT files that are generated are around 21MB in size and the ATF only allows a 12MB max file size.  Trying to figure out how to shrink the pics embedded in the EFT file.

**Reply by ALeppala** [reply]
I got the size of the EFT to be smaller by adjusting line 140 of the file named "fingerprint.py"
Adjusted to the following: "def convert(self, encoding='jp2', ratio=25):"

built again and ran.

Files are about 8MB now.  Next problem: I get the following error when trying to upload to ATF Form 1:
![errors](https://github.com/NoahGWood/OpenEFT/assets/7120459/207494da-e20e-42a6-9188-379f0903ce8c)

**Reply by LuckyDucky87** [reply]
@ALeppala I believe I did follow the instructions. It works when I use the original fd-258.py file, but I only get 3 print boxes in the EFT. Couldn't you resize the eft file by making the fd-258 scan lower quality too? I mean instead of editing the code.

As for your error from ATF website, it looks like the Type 14 record referred to in etf_helper.ph is the fingerprints. So ATF saying in only has 3 is in line with the comment Noah made that ATF only looks at 3 of the prints, not all 15. That still confuses me though if UPS printscan eft files have 14 of the 15 print boxes in them, why wouldn't those throw an error on the ATF site?

I'm not sure about the other errors other than it appears to be a subfield for the fingerprint records.

**Reply by ALeppala** [reply]
I'll retry with an original 3 pic EFT file and see what happens with the validation.  I tried playing with different dpi settings on my scanner. Didn't really change the end result file size much.

**Reply by ALeppala** [reply]
Well I'll be dipped...  Validation worked with just the 3 pics...  
![validation](https://github.com/NoahGWood/OpenEFT/assets/7120459/b766adef-2e05-4fb9-803b-cc86ae827a91)

**Reply by LuckyDucky87** [reply]
@ALeppala I guess the question is will the ATF approve a Form 1 with the 3 print EFT?

@NoahGWood

**Comment by ALeppala** [comment]
We'll find out... Submitted the Form1 today.

Sent from my mobile device so please excuse my typing mistakes.
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by LuckyDucky87** [reply]
@ALeppala 
Mine was approved today! I used the EFT that had the left and right slaps and the double thumbs print image. Did it on a trust.

**Reply by LuckyDucky87** [reply]
@ALeppala Form 1 on a trust that I created. It took about 12 days.

**Reply by ALeppala** [reply]
I filed a Form1 with a Trust on Aug 19.  My first Form1 (e-file) took 91 days to get rejected.  Stupid mistake, I uploaded an unsigned form.  Resubmitted and got approved in 7 days.  I waited 496 days for my paper Form4 to get processed, so the e-file process is much better overall.

**Reply by rwcowall** [reply]
Hey there! I made my EFT last night following the instructions and conversation here.  Today however at my FFL, we noticed the EFT was too large (~16MB but the file limit for the form is 12MB).  I'll look through the code/configs and see if there is anything I can do but do either of you have recommendations on how to get the file size down?

**Reply by rwcowall** [reply]
Got it to fit, cropped the original photo, and set it to black and white.  Eforms accepted the file, so I'll try at the FFL tomorrow!

**Comment by ALeppala** [comment]
Sweet!  Mine is still pending.  Keeping fingers crossed.
Did you file a Form1 or Form4?

Sent from Outlook<http://aka.ms/weboutlook>

**Comment by ALeppala** [comment]
Sorry for the late reply. I had to play with resolution settings to get the size right too.

Sent from my mobile device so please excuse my typing mistakes.
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by rwcowall** [reply]
Mine got approved Wednesday (4 days) and picked it up yesterday!  I'll add some of my findings and the changes I'd make probably next weekend.

**Comment by ALeppala** [comment]
That's amazing. I wish my Form 4 would clear. Seems like Trusts are taking a lot longer for some reason. No idea why.

Sent from my mobile device so please excuse my typing mistakes.
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by rwcowall** [reply]
Best of luck! To be fair, I'm not sure we fully know why individuals are going so quickly either.

**Comment by freinkelhouse** [comment]
Anyone  have a work around for this issue? 
   raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command './browser/linux/chrome.AppImage --app=http://localhost:8080/' returned non-zero exit status 126.

**Reply by somekid6** [reply]
Use another browser other than the chrome image in this project or switch to this other fork

https://github.com/Robbbbbbbbb/OpenEFT
